[Dynamo] Inline class method

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89098

Fixes: https://github.com/pytorch/pytorch/issues/93416

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan committed:
Inline class method

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -140,6 +140,11 @@ def __str__(self):
         return self.value_type
 
+    def as_python_constant(self):
+        if self.value_type is not types.MemberDescriptorType:
+            return self.value
+        return super().as_python_constant()
```
nit: special condition + affirmative check + early return reads cleaner

```suggestion
        if self.value_type is types.MemberDescriptorType:
            return super().as_python_constant()
        return self.value
```

tugsbayasgalan replied:
sounds good. Will do

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -397,7 +397,7 @@ def index_source(key):
                 value, guards=make_guards(GuardBuilder.FUNCTION_MATCH)
             )
-        elif istype(value, (type, ABCMeta)):
+        elif istype(value, (type, ABCMeta)) and not issubclass(value, torch.nn.Module):
```
does isinstance work here?

tugsbayasgalan replied:
value in this case is a class name, not instance.

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "[Dynamo] Inline class method"


Fixes: https://github.com/pytorch/torchdynamo/issues/1858

Class will be treated as UserDefinedObject which will treat forward as UserFunction. Then, it will just inline the function body into the top level module
cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
